python/orca/example/learn/tf2/lisewise/listwise_parquet.py

```diff
@@ -0,0 +1,170 @@
+    parser.add_option('--executor_cores', type=int, default=18,
+                      help='The executor core number.')
+    parser.add_option('--executor_memory', type=str, default="10g",
+                      help='The executor memory.')
+    parser.add_option('--num_executor', type=int, default=1,
+                      help='The number of executor.')
+    parser.add_option('--driver_cores', type=int, default=4,
+                      help='The driver core number.')
+    parser.add_option('--driver_memory', type=str, default="36g",
+                      help='The driver memory.')
+    parser.add_option("--data_dir", dest="data_dir", default="./ml-1m",
+                      help='The directory to ml-1m .dat files')
+
+    (options, args) = parser.parse_args(sys.argv)
+    data_dir = options.data_dir
+
+    if options.cluster_mode == "local":
+        init_orca_context("local", cores=options.executor_cores, memory=options.executor_memory,
+                          init_ray_on_spark=True)
+    elif options.cluster_mode == "yarn":
+        init_orca_context("yarn-client", cores=options.executor_cores,
+                          num_nodes=options.num_executor, memory=options.executor_memory,
+                          driver_cores=options.driver_cores, driver_memory=options.driver_memory,
+                          init_ray_on_spark=True, extra_python_lib="models.py")
+    else:
+        raise ValueError("cluster_mode should be 'local' or 'yarn', but got " + args.cluster_mode)
+
+    (options, args) = parser.parse_args(sys.argv)
+    data_dir = options.data_dir
+
+    dataset = {
+        "ratings": ['userid', 'movieid', 'rating', 'timestamp'],
+        "users": ["userid", "gender", "age", "occupation", "zip-code"],
+        "movies": ["movieid", "title", "genres"]
+    }
+
+    tbl_dict = dict()
+    for data, cols in dataset.items():
+        tbl = FeatureTable.read_csv(os.path.join(data_dir, data + ".dat"),
+                                    delimiter=":", header=False)
+        tmp_cols = tbl.columns[::2]
+        tbl = tbl.select(tmp_cols)
+        col_dict = {c[0]: c[1] for c in zip(tmp_cols, cols)}
+        tbl = tbl.rename(col_dict)
+        tbl_dict[data] = tbl
+
+    full_tbl = tbl_dict["ratings"].join(tbl_dict["movies"], "movieid") \
+        .dropna(columns=None).select(["userid", "title", "rating"])
+    # cast
+    full_tbl = full_tbl.cast(["rating"], "int")
+    full_tbl = full_tbl.cast(["userid"], "string")
+    train_tbl, test_tbl = full_tbl.random_split([0.85, 0.15], seed=1)
+
+
+    def preprocess(tbl, feature_cols):
+        col_dict = {"collect_list(" + c + ")": c + "s" for c in feature_cols}
+        tbl = tbl.group_by("userid", agg="collect_list")
+        tbl = tbl.rename(col_dict)
+        arr_len = lambda x: len(x)
+        tbl = tbl.apply("ratings", "len", arr_len, dtype="int")
+        min_len = tbl.get_stats("len", "min")["len"]
+        max_len = tbl.get_stats("len", "max")["len"]
```

Comment:
maybe `tbl.get_stats("len", ["max", "min"])`?

Reply:
Fixed in https://github.com/analytics-zoo/friesian/pull/239